### PR TITLE
Use PQS for sorted index if possible

### PR DESCRIPTION
**Pull request by vasukalariya**
# Description
Summarize the change.

Fixes #<issue-number> (link all the GitHub issues this addresses)

# Testing
Describe how you tested this code. How can the reviewers reproduce your tests?

# Checklist:
Before marking your pull request as ready for review, complete the following.

- [ ] I have self-reviewed this PR.
- [ ] I have removed all print-debugging and commented-out code that should not be merged.
- [ ] I have added sufficient comments in my code, particularly in hard-to-understand areas.
- [ ] I have formatted the code, if applicable. For Go, I have run `goimports -w .`.

### CODE REVIEW

File: pkg/segment/query/processor/searcher.go [hunk 1]
```diff
@@ -465,6 +510,75 @@ func (s *Searcher) handleSortIndexWithFilter(qsr *query.QuerySegmentRequest, lin
...
+
+	req := &structs.SegmentSearchRequest{
+		SegmentKey:          qsr.GetSegKey(),
+		VirtualTableName:    qsr.GetTableName(),
+		AllPossibleColumns:  s.queryInfo.GetColsToSearch(),
+		AllBlocksToSearch:   searchMetadataToSend,
+		BlockToValidRecNums: blockToValidRecNums,
+		SearchMetadata: &structs.SearchMetadataHolder{
+			BlockSummaries:    blkSummaries,
+			SearchTotalMemory: uint64(len(blkSummaries) * 16), // TODO: add bitset size here
```
Comment (AndrewHess) [inline pkg/segment/query/processor/searcher.go:541]: 1. I'm not sure what this TODO means
2. Should this 16 be 24 instead? The size of one BlockSummary is 8 + 8 + 2 + 6 (padding) = 24
```
type BlockSummary struct {
	HighTs   uint64
	LowTs    uint64
	RecCount uint16
}
```

File: pkg/segment/search/segsearch.go [hunk 2]
```diff
@@ -397,28 +397,43 @@ func rawSearchSingleSPQMR(multiReader *segread.MultiColSegmentReader, req *struc
...
 		if blkResults.ShouldIterateRecords(aggsHasTimeHt, isBlkFullyEncosed, blkSum.LowTs, blkSum.HighTs, false) {
-			for recNum := uint(0); recNum < numRecsInBlock; recNum++ {
-				if pqmr.DoesRecordMatch(recNum) {
+
+			var recordNums []uint16
+			if req.BlockToValidRecNums != nil {
+				records, ok := req.BlockToValidRecNums[blockNum]
+				if !ok {
+					// TODO: do this check in the caller.
+					return
```
Comment (AndrewHess) [inline pkg/segment/search/segsearch.go:406]: Should this be a `continue`, or is `return` correct?

**Review by AndrewHess — COMMENTED** [review]
Looking good

File: pkg/segment/search/segsearch.go (hunk 2 above)
Comment (vasukalariya) [reply pkg/segment/search/segsearch.go:406]: We are processing block over here, so if the records for that block are not present we should just return.

File: pkg/segment/query/processor/searcher.go (hunk 1 above)
Comment (vasukalariya) [reply pkg/segment/query/processor/searcher.go]: Sorry forgot to remove the TODO. Addressed in the latest commit.

File: pkg/segment/search/segsearch.go (hunk 2 above)
Comment (AndrewHess) [reply pkg/segment/search/segsearch.go:406]: But this is inside this loop on line 366
```
for blockNum := range filterBlockRequestsChan {
```
So couldn't we be processing multiple blocks and we'd just want to skip the current block? Maybe I'm missing something

Comment (vasukalariya) [reply pkg/segment/search/segsearch.go:406]: Oh yeah, sorry missed that.

Comment (vasukalariya) [reply pkg/segment/search/segsearch.go]: addressed in the latest commit.

**Review by AndrewHess — APPROVED** [review]

**Pull request merged**